src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager_Blocking.cs

```diff
@@ -0,0 +1,168 @@
+                                MetadataAddress,
+                                TelemetryConstants.Protocols.FirstRefresh,
+                                ex);
+
+                            throw LogHelper.LogExceptionMessage(
+                                new InvalidOperationException(
+                                    LogHelper.FormatInvariant(LogMessages.IDX20803, LogHelper.MarkAsNonPII(MetadataAddress ?? "null"), LogHelper.MarkAsNonPII(_syncAfter), LogHelper.MarkAsNonPII(ex)), ex));
+                        }
+                        else
+                        {
+                            _syncAfter = DateTimeUtil.Add(
+                                _timeProvider.GetUtcNow().UtcDateTime,
+                                AutomaticRefreshInterval < RefreshInterval ? AutomaticRefreshInterval : RefreshInterval);
+
+                            var elapsedTime = _timeProvider.GetElapsedTime(startTimestamp);
+
+                            TelemetryClient.LogConfigurationRetrievalDuration(
+                                MetadataAddress,
+                                elapsedTime,
+                                ex);
+
+                            LogHelper.LogExceptionMessage(
+                                new InvalidOperationException(
+                                    LogHelper.FormatInvariant(LogMessages.IDX20806, LogHelper.MarkAsNonPII(MetadataAddress ?? "null"), LogHelper.MarkAsNonPII(ex)), ex));
+                        }
+                    }
+                }
+
+                // Stale metadata is better than no metadata
+                if (_currentConfiguration != null)
+                    return _currentConfiguration;
+                else
+                    throw LogHelper.LogExceptionMessage(
+                        new InvalidOperationException(
+                            LogHelper.FormatInvariant(
+                                LogMessages.IDX20803,
+                                LogHelper.MarkAsNonPII(MetadataAddress ?? "null"),
+                                LogHelper.MarkAsNonPII(_syncAfter),
+                                LogHelper.MarkAsNonPII(_fetchMetadataFailure)),
+                            _fetchMetadataFailure));
+            }
+            finally
+            {
+                _refreshLock.Release();
+            }
+        }
+
+        private void RequestRefreshBlocking()
+        {
+            DateTime now = _timeProvider.GetUtcNow().UtcDateTime;
+
+            if (now >= DateTimeUtil.Add(LastRefresh, RefreshInterval) || _isFirstRefreshRequest)
+            {
+                _refreshRequested = true;
+                _syncAfter = now;
+                _isFirstRefreshRequest = false;
+            }
+        }
+
+        private void TelemetryForUpdateBlocking()
+        {
+            string updateMode;
+
+            if (_currentConfiguration is null)
+            {
+                updateMode = TelemetryConstants.Protocols.FirstRefresh;
```

Comment:
Is there value in tracking a separate set of update modes for blocking ops?

Reply:
It would help distinguish if one or the other is failing, so I think so yes.

Reply:
Could we put that in the Tag?

Reply:
Tag was a good fit. Can read it from the environment in right where we create the tags